src/dune_pkg/toolchain.ml

```diff
@@ -0,0 +1,256 @@
+open! Import
+open! Stdune
+
+module Make = struct
+  let path =
+    lazy
+      (* TODO: prefer gmake over make if it's available for compatibility with the BSDs *)
+      (match Bin.which ~path:(Env_path.path Env.initial) "make" with
+       | Some p -> p
+       | None ->
+         User_error.raise
+           ~hints:[ Pp.text "Install \"make\" with your system package manager." ]
+           [ Pp.text
+               "The program \"make\" does not appear to be installed. This program is \
+                needed to compile the ocaml toolchain."
+           ])
+  ;;
+end
+
+module Dir = struct
+  let make_and_check_dir path =
+    if not (Path.exists path) then Path.mkdir_p path;
+    if not (Path.is_directory path)
+    then
+      User_error.raise
+        [ Pp.textf "Expected %s to be a directory but it is not." (Path.to_string path) ]
+  ;;
+
+  let toolchain_base_dir () =
+    let cache_dir =
+      Xdg.create ~env:Sys.getenv_opt () |> Xdg.cache_dir |> Path.of_string
+    in
+    let d = Path.relative (Path.relative cache_dir "dune") "toolchains" in
+    make_and_check_dir d;
+    d
+  ;;
+end
+
+module Version = struct
+  type t = string
+
+  let all = [ "4.14.2"; "5.1.1" ]
+  let all_by_string = List.map all ~f:(fun t -> t, t)
+  let to_string t = t
+
+  let toolchain_dir t =
+    let d = Path.relative (Dir.toolchain_base_dir ()) (to_string t) in
+    Dir.make_and_check_dir d;
```

Comment:
`make_and_check_dir` is already called by `toolchain_base_dir` and while it doesn't hurt, it also doesn't help.

Maybe `make_and_check_dir` could just get inlined into `toolchain_base_dir`?